### DISCUSSION TITLE
Clone cell deletion process

### DISCUSSION THREAD

**Discussion by jost-s** [holochain/holochain#2961] [Ideas]
# Background

Clone cells can be created, enabled and disabled, as well deleted. When a clone cell is deleted, all of its data is erased and cannot be restored. Further it cannot be re-created identically, because it would lead to a source chain fork from the peers' perspective. The data is not guaranteed to be available in its entirety and private entries are never stored on the DHT, so the previous state of a cell cannot be reestablished once it's been deleted.

Every active cell causes network traffic and possibly increasing disk usage. Therefore clone cells can be disabled to prevent these effects from happening.

# Suggestion

On top of that we can extend the delete feature to provide an additional stage of deletion. Usually public entries will take up the most space of a source chain. The intermediate step would be to only delete those public entries and retain the private ones. In theory that would also make re-creation of the same clone cell possible.

In a final deletion step, private entries of the clone cell would be deleted too.

**Comment by mattyg** [comment]
I would like to hear more about the original use case(s) that prompted this discussion @matthme @zippy 

**Overall I see 3 potentially good features:**
1. Add a boolean property to dna manifest `can_delete_clones_via_app_api` and expose the `delete_clone()` function over the app api. If that property is true, cloned cells can be deleted via the app api, if not, deletion fails with an error.
2. Add a new function `request_delete_clone()`, and expose over the app api. Upon receiving this call, the conductor emits a signal over the admin api. The launcher needs to listen for this signal and take some action (i.e. expose a dialog to the user to confirm deletion).
3. Add 2 additional boolean parameters to the `disable_clone()` function: `delete_public_entries` and `delete_private_entries`. If set to true, entry data will be deleted after the cell is disabled, as specified.

**Reply by maackle** [reply]
I think this is all reasonable, but I think (3) should be a separate call (dehydrate_clone?) which deletes entries and disables the cell, and which takes a param specifying whether to delete only public entries or all entries

**Reply by matthme** [reply]
Regarding use cases: Any use case where you want to have the ability to free up disk space with a certain degree of granularity. For example a file sharing app that may take up significant amount of disk space but you don't want to delete the app as a whole if you reach the limits of your hard drive. Or "throw-away DHT's" that have a certain functionality for a certain period of time and then should not further take up any disk space once they fulfilled their purpose.

**Comment by nphias** [comment]
i like the notion of dehydrating cells.. and auto-disable archive options
The usecase i had in mind was privacy related and perhaps a 4th feature.. spin up a temporary DHT where once all parties have finished coordinating the cell copies self-destruct.
 The parties involved all agree and understand the ephemeral nature of the communication.. it could be done with a TTL of a number of hours.   Yes its true that this removes agency from the cell cloners. But it addresses certain usecases backed up by contractual/legal agreements where data is not meant to be stored.  Of course the cell cloner can hack the code/conductor in a way where the cell doesnt get deleted, but thats not the point, legal documents/ramifications akin to GDPR disincentivise such actions

**Reply by maackle** [reply]
Maybe we could make `delete_clone` available from the HDK so that it can be set as a scheduled function, I think that would accomplish what you're looking for? If not that, even just having the app UI issue the deletion at the right time, could get us most of the way there.
